### PR TITLE
[PropertyInfo] consider write property visibility to decide whether a property is writable

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | part of #58556
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 1]
```diff
@@ -621,6 +621,10 @@ private function isAllowedProperty(string $class, string $property, bool $writeA
                 return false;
             }
 
+            if (\PHP_VERSION_ID >= 80400 && ($reflectionProperty->isProtectedSet() || $reflectionProperty->isPrivateSet())) {
```
Comment (pan93412) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: `$writeAccessRequired` should be satisfied before checking if setter is protected or private?

Besides, how about using `method_exists` to make Psalm happy?

**Review by pan93412 — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/ReflectionExtractorTest.php [hunk 2]
```diff
@@ -685,4 +686,17 @@ public static function extractConstructorTypesProvider(): array
             ['ddd', null],
         ];
     }
+
+    /**
+     * @requires PHP 8.4
+     */
+    public function testAsymetricVisibility()
```
Comment (pan93412) [inline src/Symfony/Component/PropertyInfo/Tests/Extractor/ReflectionExtractorTest.php]:
```suggestion
    public function testAsymmetricVisibility()
```

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: checking for the `PHP_VERSION_ID` constant can be done at compile time, the same is not true when using `method_exists()` instead

**Review by pan93412 — APPROVED** [review]

**Review by alexandre-daubois — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**